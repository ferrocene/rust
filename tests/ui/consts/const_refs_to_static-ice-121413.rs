// ICE: ImmTy { imm: Scalar(alloc1), ty: *const dyn Sync } input to a fat-to-thin cast (*const dyn Sync -> *const usize
// or with -Zextra-const-ub-checks: expected wide pointer extra data (e.g. slice length or trait object vtable)
// issue: rust-lang/rust#121413
//@ compile-flags: -Zextra-const-ub-checks
// ignore-tidy-linelength
const REF_INTERIOR_MUT: &usize = {
    //~^ HELP consider importing this struct
    static FOO: dyn Sync = AtomicUsize::new(0);
    //~^ ERROR failed to resolve: use of undeclared type `AtomicUsize`
    //~| ERROR the size for values of type `(dyn Sync + 'static)` cannot be known at compilation time
    //~| HELP the trait `Sized` is not implemented for `(dyn Sync + 'static)`
    unsafe { &*(&FOO as *const _ as *const usize) }
};
pub fn main() {}
