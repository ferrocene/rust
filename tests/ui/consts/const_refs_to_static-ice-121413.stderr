error[E0433]: failed to resolve: use of undeclared type `AtomicUsize`
  --> $DIR/const_refs_to_static-ice-121413.rs:8:28
   |
LL |     static FOO: dyn Sync = AtomicUsize::new(0);
   |                            ^^^^^^^^^^^ use of undeclared type `AtomicUsize`
   |
help: consider importing this struct
   |
LL + use std::sync::atomic::AtomicUsize;
   |

error[E0277]: the size for values of type `(dyn Sync + 'static)` cannot be known at compilation time
  --> $DIR/const_refs_to_static-ice-121413.rs:8:17
   |
LL |     static FOO: dyn Sync = AtomicUsize::new(0);
   |                 ^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `(dyn Sync + 'static)`
   = note: statics and constants must have a statically known size

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0433.
For more information about an error, try `rustc --explain E0277`.
